### PR TITLE
Login: Handle Jetpack site connection for self-hosted sites

### LINKED ISSUE #7846: Login: Handle Jetpack connection for self-hosted sites without site-level connection
(A clone of https://github.com/woocommerce/woocommerce-android/issues/7525)

Jetpack connection have two stages connection:
1. Site level connection that gets setup on the initial setup of Jetpack, at this step, just some Jetpack features are available.
2. Account level connection that gets setup once the user connects the site to their WordPress.com account.

The first part is done for each site just once, while the second part can be done for each account separately, and Jetpack features won't be fully available until the second step is done. (source: PCYsg-kqV-p2).

When working on #7597, we implemented Jetpack Connection, but we found out that our logic was working only for the site's that have already did the step 1, this issue is tracking the implementation for sites without the site-level connection too.

### Technical details:
To handle the connection, the apps currently call the Jetpack endpoint [`/jetpack/v4/connection/url`](https://github.com/Automattic/jetpack/blob/trunk/docs/rest-api.md#get-wp-jsonjetpackv4connection), this API returns a different response depending on the state of the Jetpack connection on the site:
- If the site didn't implement the site-level connection, the URL would be at the form: `https://{site_url}/wp-admin/admin.php?page=jetpack&action=register&_wpnonce={nonce}`.
- If the site has at least a site-level connection, the URL would directly point to WordPress.com servers: `https://jetpack.wordpress.com/jetpack.authorize/1/...`

The constraint with the first URL, is that to allow the connection, it has to use the same authentication Cookies as the ones used to retrieve it, and when it comes to our apps, this means two approaches:
1. Syncing the Cookies between the API network client and the WebView used for the connection.
2. The alternative way would be to call the connection URL using the same network client as the API, while ensuring automatic redirection is disabled, then we can retrieve the WordPress.com connection URL using the header `Location`.

### UI details
We just need to ensure that the WordPress.com ToS are mentioned in the screen that initiates the connection.

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes #7846 

### Description

When working on #7597 and #7745, we added Jetpack connection handling for accounts without connection to Jetpack sites. 

There's another case where we need to handle Jetpack connection - when a site has not registered with Jetpack. This is when we need to handle site connection before proceeding to account connection.

To successfully register Jetpack connection for a site, we need to request the connection URL, then access it with the same cookies used to request the URL. Otherwise, the URL will be considered expired.

The solution introduced in this PR includes:
- Only show Jetpack error if a user tries to log in to a site that does not have Jetpack installed or activated. If the site only misses Jetpack connection, show them the account mismatch screen, where we can help them register the Jetpack connection on their site.
- Retrieve cookies from WordPressOrgNetwork
- Send retrieved cookies to AuthenticatedWebViewController
- Add a link to show terms and conditions to be agreed on when connecting Jetpack.
- Also: update the primary button title on the Account mismatch screen to "Connect Jetpack to your account" to align with Android.

### Testing instructions

Handling Jetpack connection when logging in with site credentials:
- Pre-requisite: create a JN site with Jetpack installed, activated but not connected. The site may or may not have Woo.
- Log out of the app if needed. On the prologue screen, select Enter site address.
- Enter the address of the JN site.
- Select Log in with site credentials.
- Enter the credentials and tap Continue. Notice that the account mismatch screen is then displayed.
- Notice a new text at the bottom of the screen regarding the terms to be agreed upon when connecting Jetpack.
- Tap Connect Jetpack to your account. Notice that a web view is displayed navigating to the Jetpack authorization screen. Tap Authorize, notice that the app will then navigate to the WP.com login flow and show the home screen after authentication succeeds.

Handling Jetpack connection when logging in with a WP.com account:
- Pre-requisite: create a JN site with Jetpack installed, activated but not connected. The site may or may not have Woo.
- Log out of the app if needed. On the prologue screen, select Enter site address.
- Enter the address of the JN site.
- Enter WP.com credentials and tap Continue. Notice that the account mismatch screen is then displayed.
- Notice a new text at the bottom of the screen regarding the terms to be agreed upon when connecting Jetpack.
- Tap Connect Jetpack to your account. Notice that the site credential login flow is presented. Enter the credentials and tap Continue.
- Notice that a web view is displayed navigating to the Jetpack authorization screen. Tap Authorize, notice that the app will then navigate to the WP.com login flow and show the home screen after authentication succeeds.

Please feel free to test with a mismatch WP.com site. Notice that there is no terms text at the bottom of the account mismatch screen.

### Screenshots

https://user-images.githubusercontent.com/5533851/195795881-12fe325a-fd53-4d37-a910-45837241d9e8.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift [hunk 1]
```diff
@@ -128,9 +133,14 @@ private extension AuthenticatedWebViewController {
...
         }
-        if let credentials = ServiceLocator.stores.sessionManager.defaultCredentials {
+        /// Only authenticate for WP.com automatically if no cookies are provided.
+        ///
+        if let credentials = ServiceLocator.stores.sessionManager.defaultCredentials, cookies.isEmpty {
             webView.authenticateForWPComAndRedirect(to: url, credentials: credentials)
         } else {
+            for cookie in cookies {
+                webView.configuration.websiteDataStore.httpCookieStore.setCookie(cookie)
+            }
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift]: ❓  Is this for automatically logging into the site using .org credentials? 

Would it make sense to use a helper method like `func setCookiesToAuthenticateForDotOrgSignIn()` to explain this? (Sorry, I can't come up with a better name. 😄 )

File: WooCommerce/Classes/Authentication/Navigation Exceptions/WrongAccountErrorViewModel.swift [hunk 2]
```diff
@@ -117,6 +123,32 @@ final class WrongAccountErrorViewModel: ULAccountMismatchViewModel {
                 self?.analytics.track(event: .LoginJetpackConnection.jetpackConnectionErrorShown(selfHostedSite: isSelfHosted))
             }
 
+        $isSelfHostedSite
+            .map { [weak self] isSelfHosted -> NSMutableAttributedString in
+                // only shows terms text if the site is self-hosted,
+                // since the user cannot handle Jetpack connection themselves on WP.com sites.
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Navigation Exceptions/WrongAccountErrorViewModel.swift]: Thanks for adding comments. 👏

File: WooCommerce/Classes/Authentication/Navigation Exceptions/WrongAccountErrorViewModel.swift [hunk 3]
```diff
@@ -117,6 +123,32 @@ final class WrongAccountErrorViewModel: ULAccountMismatchViewModel {
                 self?.analytics.track(event: .LoginJetpackConnection.jetpackConnectionErrorShown(selfHostedSite: isSelfHosted))
             }
 
+        $isSelfHostedSite
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Navigation Exceptions/WrongAccountErrorViewModel.swift]: Super nit: Would it make sense to move this out of `viewDidLoad` to reduce the number of lines in `viewDidLoad`?

**Review by selanthiraiyan — APPROVED** [review]
**Testing**
It works as expected ✅ 

**Code**
Left a few non-blocking comments/questions. 

LGTM 🚢 🎉

**Comment by selanthiraiyan** [comment]
>Please feel free to test with a mismatch WP.com site. Notice that there is no terms text at the bottom of the account mismatch screen.

I tested this for a self-hosted site (`https://genuine-barnacle.jurassic.ninja/`) and a WPCOM account which isn't associated with that site. 

i.e. `https://public-api.wordpress.com/rest/v1.1/me/sites` response for that WPCOM account doesn't have `https://genuine-barnacle.jurassic.ninja/` in it. 

But I still see the terms text at the bottom. Am I missing something here?

**Comment by itsmeichigo** [comment]
Thank you @selanthiraiyan for the review!

> But I still see the terms text at the bottom. Am I missing something here?

This is intentional. So the expectation is that the terms text is only available for self-hosted sites because we can help users connect Jetpack on their site - and not for WP.com sites because they need an invitation from a site admin to be connected.

So to test that case about the mismatch WP.com site, you need to log in to a WP.com (Atomic) site with a mismatch WP.com account instead. The terms text should not be available then.

**Another update**: I've added a few more commits to restore the automatic WP.com authentication in `AuthenticatedWebViewController` when handling Jetpack site connection.

- Previously, I retrieved the cookies from `WordPressOrgNetwork` and injected them into the `AuthenticatedWebViewController`:   
  - The cookies make it possible to handle the site connection URL
  - However, they conflict with the automatic WP.com authentication, so I added a workaround to remove the authentication if there are injected cookies. 
  - This means that users will have to enter their WP.com credentials again in the web view even if they already did in the login flow earlier.
 
- Then, I followed Hicham's suggestion [p1665912798859479/1665502423.023859-slack-C03L1NF1EA3]:
  - Removed the injection of cookies to  `AuthenticatedWebViewController`.
  - Added a read-only variable `session` to `Network`. This is needed to retrieve cookies from the session of the network client.
  - Updated `JetpackConnectionNetwork` to handle Jetpack site connection registration. I used `URLSession` for simplicity to handle redirection; the idea is to inject the necessary cookies and request the site connection URL. 
  - When the site connection URL is triggered, a redirection will be requested to navigate to the account connection URL. We want to show this URL in a web view, so the redirection should be canceled (using `URLSessionDataDelegate`). 
  - The redirected URL will then be saved and sent to the `JetpackConnectionStore`. 
  - A web view will then be displayed and show the account connection URL, with the WP.com authentication already handled. If the WP.com account has 2FA enabled, a code may still be requested (see the recording below).


https://user-images.githubusercontent.com/5533851/196150289-9ed046a9-a8d0-4208-ad0f-600ab9118c84.mp4

I know this is all very complicated 😅 Please feel free to ping me with any questions you have, I'll try my best to explain.

I haven't added new unit tests for the new updates; I'm not sure what I can do about that. Please also leave any suggestions if necessary.

**Comment by selanthiraiyan** [comment]
>This is intentional. So the expectation is that the terms text is only available for self-hosted sites because we can help users connect Jetpack on their site - and not for WP.com sites because they need an invitation from a site admin to be connected.

I understand. Thanks for explaining this.

**Comment by selanthiraiyan** [comment]
Thanks for the detailed explanation about the new updates, @itsmeichigo 

This still LGTM. 🎉

**Review by selanthiraiyan — APPROVED** [review]

**Pull request merged**